### PR TITLE
Adding Multi-DC Cassandra Support to Bootstrap Configuration Script(s)

### PR DESCRIPTION
**Pull request by Kerptastic**
## Which problem is this PR solving?
- Solves: https://github.com/jaegertracing/jaeger/issues/2618

TL;DR from the issue:
- Multi-DC Cassandra setups are not supported by the jaeger-cassandra bootstrap scripts (i.e., can only provide a single Datacenter value, versus `N` Datacenter values, where `N > 1`)
- Even numbered quorum value is being used for `prod` setups, versus using an odd number

## Short description of the changes
Adding the capability to provide a comma-delimited list of data-centers from the jaeger-operator, through `docker.sh`, and into `create.sh` to allow for multi-DC Cassandra setups.
- Added internal field separator to `create.sh`
- Set default replication factor for prod setups to be 3, and provide minimum odd-number quorum
- Create replication factor string for 1+ Datacenter values when in prod mode
- Updated `echo` statements accordingly

### CODE REVIEW

File: plugin/storage/cassandra/schema/create.sh [hunk 1]
```diff
@@ -30,9 +31,19 @@ if [[ "$MODE" == "" ]]; then
     usage "missing MODE parameter"
 elif [[ "$MODE" == "prod" ]]; then
     if [[ "$DATACENTER" == "" ]]; then usage "missing DATACENTER parameter for prod mode"; fi
+    replication_factor=${REPLICATION_FACTOR:-3}
```
Comment (joe-elliott) [inline plugin/storage/cassandra/schema/create.sh:34]: You should be able to just set the env var `REPLICATION_FACTOR=3` on the container running this script.  I don't think we should change the default value here.

Comment (Kerptastic) [reply plugin/storage/cassandra/schema/create.sh:34]: This is sort of the can of worms I was poking at in the issue, and understand the hesitancy to change to 3. I am not a Cassandra expert by any stretch of imagination, but was informed internally that for quorum, 3 should be the default vs. 2.

Also, if you are running `create.sh` manually or within self baked images, that would work fine, but the issue is that the `docker.sh` (lines 42-46) is where the script is invoked, which is provided by the `jaeger-cassandra-schema` image. From what I understand, folks using the Jaeger Operator Helm Chart(like ourselves), we do not have a way to push in the replication factor via `values.yaml`.

For Example:
```
...
        cassandraCreateSchema:
          enabled: true
          datacenter: us-west2,europe-west4
          mode: "prod"
          image: my-custom-images/jaeger-cassandra-schema-bootstrap:latest
...
```

Or am I overthinking this?

Reference link to `docker.sh`:
https://github.com/jaegertracing/jaeger/blob/44a6d37edeeabb669dc1df61a2099bed09c1bb9a/plugin/storage/cassandra/schema/docker.sh#L45

Comment (yurishkuro) [reply plugin/storage/cassandra/schema/create.sh:34]: There is no "right" value for replication factor, there are pros and cons for different settings. I agree that for backwards compatibility it's best not to change the current default.

Regarding the config schema above, it sounds like it needs a change in the operator.

File: plugin/storage/cassandra/schema/create.sh [hunk 2]
```diff
@@ -30,9 +31,19 @@ if [[ "$MODE" == "" ]]; then
...
+    count=0
+    for dc in $DATACENTER; do
+      if [[ "$count" -eq "0" ]]; then
+        topology_config="'$dc': '${replication_factor}'"
+      else
+        topology_config="$topology_config, '$dc': '${replication_factor}'"
+      fi
+      count=$((count+1))
+    done
+    replication="{'class': 'NetworkTopologyStrategy', ${topology_config}}"
```
Comment (joe-elliott) [inline plugin/storage/cassandra/schema/create.sh:46]: For more complex multi dc replication strategies does it make sense to just allow the operator to override the entire replication string using an environment variable? 

If we add this then eventually someone will want datacenters with different RF values.

File: plugin/storage/cassandra/schema/create.sh (hunk 1 above)
Comment (Kerptastic) [reply plugin/storage/cassandra/schema/create.sh:34]: @yurishkuro I agree on both fronts. I think making the config more flexible at the operator level is ultimately the right solution.

File: plugin/storage/cassandra/schema/create.sh (hunk 2 above)
Comment (Kerptastic) [reply plugin/storage/cassandra/schema/create.sh:46]: @joe-elliott @yurishkuro I agree. I think having the entire string would be simple from an implementation strategy, but possibly a bit harder to debug errors for users who are have little Cassandra experience, or deploying at the operator. I think there are multiple options here:

1. Provide the entire replication config - this would be the easiest to implement, but would be pretty awful to put into a yaml file for the operator config: 
```
{ 'class': 'NetworkTopologyStrategy', 'dc1': '1', 'dc2': '2' ... }
```
2. Provide just the data-center portion, and still use the `mode` setting - this is less offensive to a user of the operator, and pretty simple to validate/implement:
```
'dc1': '1', 'dc2': '2', ...
```
3. Provide a parseable format, such as DC1:RF1,DC2:RF2 - probably the simplest on all fronts, development and using it in yaml, but it would require informing the users specifically what it is:
```
dc1:1,dc2:2,...
```
I would suggest option 3, but I come from the inexperienced with Cassandra camp, and don't know if this would be confusing and hard to understand. I don't fully believe so. Another negative, though, would be losing backwards compatibility with previous versions of the operator, as I would expect a `replicationFactorConfig` value of sorts to be introduced. Obviously there are ways to provide both and bark at the user if both are provided, but I will leave that up for discussion.

Let me know what y'all think!

Comment (yurishkuro) [reply plugin/storage/cassandra/schema/create.sh:46]: I like the idea of just providing a full replacement string for replication config, instead of building complicated piecemeal support.

Comment (Kerptastic) [reply plugin/storage/cassandra/schema/create.sh:46]: @yurishkuro @joe-elliott Alright! Just want to be clear, that preference is for Option 1 above? I can update this PR and start getting to work on the Operator changes. I may need some help on that side, as my golang is pretty elementary, but I'm sure I can get things started.

Comment (yurishkuro) [reply plugin/storage/cassandra/schema/create.sh:46]: Yes, let's do option 1.

Comment (Kerptastic) [reply plugin/storage/cassandra/schema/create.sh:46]: OK - my goal will be to leave the `$DATACENTER` usage as previously provided, and add support for `$REPLICATION_FACTOR_CONFIG`, that when provided, will override the creation of the `replication_factor` with the provided value. Will also PR the Operator. Thanks for the input guys, will circle back over the next few days as I get time to work on this.

**Comment by yurishkuro** [comment]
@Kerptastic I am going to close this, feel free to reopen or create another PR with the new changes.

**Comment by LuChenjing** [comment]
Hi @Kerptastic @yurishkuro ,I met the same issue... May I ask when this PR will been merged?

**Comment by Kerptastic** [comment]
> Hi @Kerptastic @yurishkuro ,I met the same issue... May I ask when this PR will been merged?

Haven't gotten around to finishing this up, and also needed to make a change to the Jaeger operator CRD to take the configuration into account.

**Comment by LuChenjing** [comment]
> > Hi @Kerptastic @yurishkuro ,I met the same issue... May I ask when this PR will been merged?
> 
> Haven't gotten around to finishing this up, and also needed to make a change to the Jaeger operator CRD to take the configuration into account.
Okay，understand, thanks.
